crates/uplc/src/optimize.rs

```diff
@@ -2,27 +2,42 @@ use crate::ast::{Name, Program};
 
 pub mod interner;
 pub mod shrinker;
+pub mod shrinker2;
 
 pub fn aiken_optimize_and_intern(program: Program<Name>) -> Program<Name> {
-    program
-        .inline_constr_ops()
-        .bls381_compressor()
-        .builtin_force_reducer()
-        .lambda_reducer()
-        .inline_reducer()
-        .identity_reducer()
-        .lambda_reducer()
-        .inline_reducer()
-        .force_delay_reducer()
-        .cast_data_reducer()
-        .builtin_eval_reducer()
-        .convert_arithmetic_ops()
-        .builtin_curry_reducer()
-        .lambda_reducer()
-        .inline_reducer()
-        .identity_reducer()
+    let mut prog = program.run_once_pass();
+
+    let mut prev_count = 0;
+
+    loop {
+        let (current_program, context) = prog.multi_pass();
+
+        if context.node_count == prev_count {
+            prog = current_program;
+            break;
+        } else {
+            prog = current_program;
+            prev_count = context.node_count;
+        }
+    }
+
+    prog = prog
         .builtin_curry_reducer()
-        .lambda_reducer()
-        .inline_reducer()
-        .remove_no_inlines()
+        .multi_pass()
+        .0
+        .builtin_curry_reducer();
+
+    loop {
+        let (current_program, context) = prog.multi_pass();
+
+        if context.node_count == prev_count {
+            prog = current_program;
+            break;
+        } else {
+            prog = current_program;
+            prev_count = context.node_count;
+        }
+    }
```

Comment:
Is this guaranteed to always converge to a fixed point though? I could imagine infinite cycles with an optimization pass adding some nodes, and the next one removing them. So we might need to track not the previous count, but the `min(all previous counts)`.

Reply:
All the optimizations in the multi pass remove terms. Except for Lambda Reducer where there is the rare exception where var that holds delay error can turn into 2 terms i.e. (delay (error )). But that is a dead end term that cannot be further affected. The main thing is that the optimizations that can create lambdas (curry, bls_hoisting, and force wrapping) are all run a fixed amount of times to prevent infinite looping.